node/smeshing_service_test.go

```diff
@@ -0,0 +1,240 @@
+	cfg.SMESHING.CoinbaseAccount = types.GenerateAddress([]byte{1}).StringWithHRP(cfg.NetworkHRP)
+	cfg.SMESHING.Opts.DataDir = filepath.Join(tmp, "post")
+	cfg.SMESHING.Opts.NumUnits = cfg.POST.MinNumUnits + 1
+	cfg.SMESHING.Opts.Scrypt.N = 2
+	cfg.SMESHING.Opts.ProviderID.SetUint32(initialization.CPUProviderID())
+
+	cfg.HARE3.RoundDuration = 2
+	cfg.HARE3.PreroundDelay = 1
+
+	cfg.HARE4.RoundDuration = 2
+	cfg.HARE4.PreroundDelay = 1
+
+	cfg.LayerAvgSize = 5
+	cfg.LayersPerEpoch = 3
+	cfg.TxsPerProposal = 100
+	cfg.Tortoise.Zdist = 5
+
+	cfg.HareEligibility.ConfidenceParam = 1
+
+	cfg.Genesis = config.DefaultTestGenesisConfig()
+	cfg.POSTService = activation.DefaultTestPostServiceConfig()
+
+	return &cfg
+}
+
+func TestNewSmeshingService(t *testing.T) {
+	t.Run("creates new service with valid config", func(t *testing.T) {
+		cfg := getSmeshingServiceTestConfig(t)
+		service, err := NewSmeshingService(cfg, zaptest.NewLogger(t))
+		require.NoError(t, err)
+		require.NotNil(t, service)
+
+		// Verify DataDir was created
+		dataDir := cfg.DataDir()
+		_, err = os.Stat(dataDir)
+		require.NoError(t, err, "DataDir should exist")
+	})
+
+	t.Run("creates identity if not exists", func(t *testing.T) {
+		cfg := getSmeshingServiceTestConfig(t)
+		service, err := NewSmeshingService(cfg, zaptest.NewLogger(t))
+		require.NoError(t, err)
+		require.NotNil(t, service)
+		require.Len(t, service.signers, 1)
+
+		// Verify identity file was created
+		_, err = os.Stat(filepath.Join(cfg.DataDir(), "identities", "local.key"))
+		require.NoError(t, err)
+	})
+}
+
+func TestSmeshingService_Start(t *testing.T) {
+	t.Run("starts API services", func(t *testing.T) {
+		cfg := getSmeshingServiceTestConfig(t)
+		cfg.API.PublicServices = []grpcserver.Service{grpcserver.Debug}
+
+		service, err := NewSmeshingService(cfg, zaptest.NewLogger(t))
+		require.NoError(t, err)
+
+		ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+		defer cancel()
+
+		errCh := make(chan error)
+		go func() {
+			errCh <- service.Start(ctx)
+		}()
+		select {
+		case <-service.started:
+		case <-ctx.Done():
+			t.Fatal("not started on time")
+		}
+
+		conn, err := grpc.NewClient(
+			service.grpcPublicServer.BoundAddress,
+			grpc.WithTransportCredentials(insecure.NewCredentials()),
+		)
+		require.NoError(t, err)
+		t.Cleanup(func() { require.NoError(t, conn.Close()) })
+
+		client := pb.NewDebugServiceClient(conn)
+		resp, err := client.ChangeLogLevel(ctx, &pb.ChangeLogLevelRequest{
+			Module: "poet",
+			Level:  "DEBUG",
+		})
+		require.NoError(t, err)
+		require.NotNil(t, resp)
+	})
+}
+
+func TestSmeshingService_StartSmeshing(t *testing.T) {
+	t.Run("starts smeshing with valid coinbase", func(t *testing.T) {
+		cfg := getSmeshingServiceTestConfig(t)
+		service, err := NewSmeshingService(cfg, zaptest.NewLogger(t))
+		require.NoError(t, err)
+
+		ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+		defer cancel()
+
+		err = service.startServices(ctx)
```

Comment:
Wait... Is `startServices` not blocking? Then it shouldn't take a `context.Context` otherwise this can lead to issues that are hard to find.

Reply:
Unfortunately, some of the services in the codebase start() are blocking, others are non-blocking. This start/stop follows exactly implementation in the `App` because the point of this PR is not refactoring the whole app lifetime handling (and refactoring half of the services in the process). The point is to strip out everything not needed for the smashing service and have a clear view what dependency it has and that it doesn't create/start services that aren't required.

Reply:
I am not sure how to continue with this, this should have been fixed on the node side a long time ago and I'm kind of hesitant to build more incorrect code on top of a shaky base here 😕